libcontainer/nsenter/cloned_binary.c

```diff
@@ -394,6 +400,99 @@ static int seal_execfd(int *fd, int fdtype)
 	return -1;
 }
 
+struct bindfd_child_args {
+	int sockfd;
+	const char *mount_target;
+};
+
+static int bindfd_in_subprocess(void *arg)
+{
+	/*
+	 * In the interests of efficiency (read: minimizing the syscall count)
+	 * and conciseness, no attempt is made to release resources which would
+	 * be cleaned up automatically on process exit, i.e. when this function
+	 * returns. This includes filesystem mounts, as this function is
+	 * executed in a dedicated mount namespace.
+	 */
+
+	/*
+	 * For obvious reasons this won't work in rootless mode because we
+	 * haven't created a userns -- but getting that to work will be a bit
+	 * complicated and it's only worth doing if someone actually needs it.
+	 */
+	if (mount("none", "/", NULL, MS_SLAVE | MS_REC, NULL) < 0)
+		return errno;
+	/*
+	 * The kernel refuses to bind-mount from the magic symlink when the
+	 * process has been cloned (or unshared) into a new mount namespace,
+	 * at least on Linux 4.4.
+	 */
+	char linkbuf[PATH_MAX + 1] = { 0 };
+	ssize_t linkpathlen = readlink("/proc/self/exe", linkbuf, sizeof(linkbuf));
```

Comment:
If we cannot use `/proc/self/exe` (or `/proc/$runc_pid/exe`) then we can't really do this. It's not safe to be operating on random paths that you get from resolving `/proc/self/exe` (and more importantly, this won't be doable once we start work on porting runc to libpathrs). (Obviously this is running "on the host" but I just don't feel comfortable with resolving magiclinks because they aren't symlinks at all and any similarity to symlinks is a very leaky illusion.)

However, I'm not sure it's actually true that you cannot use the magiclink as a mount source -- based on a quick test with `unshare` (Linux 5.19.10) it works without issue and I'm not sure what would've caused it to fail in older kernel versions (`nd_jump_link` really doesn't care about what you're doing with the path and the mount infrastructure doesn't care about the source IIRC). You probably cannot use it as a mount *target* but that shouldn't be an issue. What problem did you run into when using it as a mount source?

Reply:
On both Linux 4.4.0-1122-aws and Linux 5.15.0-33-generic (Ubuntu) any attempts to bind-mount `/proc/self/exe` in an unshared mount namespace yield `EINVAL`.
```console
ubuntu@ubuntu2204:~$ sudo strace -f -e trace=mount ./runc run --bundle bundle foobar
strace: Process 62293 attached
strace: Process 62292 attached
strace: Process 62294 attached
strace: Process 62295 attached
strace: Process 62296 attached
strace: Process 62298 attached
[pid 62298] mount("none", "/", NULL, MS_REC|MS_SLAVE, NULL) = 0
[pid 62298] mount("/proc/self/exe", "/run/runc/foobar/runc.G52pod", 0x563c345027e2, MS_BIND, 0x563c345027e2) = -1 EINVAL (Invalid argument)
[pid 62298] +++ exited with 22 +++
[...]
```

`mount(8)` canonicalizes the source path with `realpath(3)` which is why it appeared to work in your test.
```console
ubuntu@ubuntu2204:~$ sudo unshare --mount
root@ubuntu2204:/home/ubuntu# strace -e trace=readlink,mount mount --bind /proc/self/exe selfexe
readlink("/proc", 0x7fff7a3b6420, 1023) = -1 EINVAL (Invalid argument)
readlink("/proc/self", "62403", 1023)   = 5
readlink("/proc/62403", 0x7fff7a3b6420, 1023) = -1 EINVAL (Invalid argument)
readlink("/proc/62403/exe", "/usr/bin/mount", 1023) = 14
readlink("/usr", 0x7fff7a3b6420, 1023)  = -1 EINVAL (Invalid argument)
readlink("/usr/bin", 0x7fff7a3b6420, 1023) = -1 EINVAL (Invalid argument)
readlink("/usr/bin/mount", 0x7fff7a3b6420, 1023) = -1 EINVAL (Invalid argument)
readlink("/home/ubuntu/selfexe", 0x7fff7a3b6420, 1023) = -1 EINVAL (Invalid argument)
mount("/usr/bin/mount", "/home/ubuntu/selfexe", 0x563f8a351840, MS_BIND, NULL) = 0
+++ exited with 0 +++
```

<details><summary>Minimal reproducer:</summary>

```c
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <sched.h>
#include <sys/mount.h>

int main(int argc, char** argv) {
	if (argc > 1) {
		puts("Unsharing mount namespace...");
		if (unshare(CLONE_NEWNS) < 0) {
			perror("unshare");
			exit(EXIT_FAILURE);
		}
	}
	if (mount("/proc/self/exe", "selfexe", "", MS_BIND, "") < 0) {
		perror("mount");
		exit(EXIT_FAILURE);
	}
	int ret = system("/bin/sh");
	if (ret < 0) {
		perror("system");
		exit(EXIT_FAILURE);
	}
	return ret;
}
```

```console
$ gcc main.c && touch selfexe && sudo ./a.out unshare
Unsharing mount namespace...
mount: Invalid argument
$ sudo ./a.out
# exit
$ sudo umount selfexe
```

</details>

Indirecting the mount through `/proc/self/fd/<n>` also yields `EINVAL`, regardless of whether the file descriptor to `/proc/self/exe` was opened before or after the mount namespace was unshared, with or without `O_PATH`. My best guess of what's happening is that `/proc/self/exe` resolves to a `struct path` value which, being magic, has its `mnt->mnt_ns` set to the parent mount namespace, [which trips the `check_mnt()` validation here.](https://elixir.bootlin.com/linux/v6.0/source/fs/namespace.c#L2399) If my interpretation is correct, there is no way to successfully bind-mount a process' executable across namespaces without going through path resolution machinery in the process' current mount namespace.

Would you feel more comfortable if there was more assurance that the path resolved to the same file as `/proc/self/exe` @cyphar? Here's what I'm thinking:
1. `open(O_PATH)` both `/proc/self/exe` and `realpath("/proc/self/exe")`
2. Check that the file descriptors refer to the same file by `fstat`'ing both and testing whether they have the same `st_dev` and `st_ino` values
3. Bind-mount the `/proc/self/fd/<n>` for the realpath fd

Reply:
Yes, you were both quite right -- I had forgotten about the annoying way `mount(8)` handles paths. The future libpathrs issue is still a problem but I think we need to drop this bindfd code before then anyway, so we can cross that bridge if we ever get to it.